### PR TITLE
Add the Flex component and variant support

### PR DESCRIPTION
**Pull request by wesleycole**
Creates the `Flex` primitive component which is a wrapper for the `Box` component and adds variant support to `Box`.

**Main Story:** [ch28154](https://app.clubhouse.io/jilt/story/28154/create-flexbox-examples-in-storybook-using-the-flex-component)

<a name="details"></a>

## Details

This creates a wrapping component on Box that adds the `display: flex` property. 

The majority of this work is adding the `variant` prop to the Box component. This will allow all components that wrap `Box` and for `Box` itself to be styled by variants. This will allow developers to create variants in the theme like:

```
// theme.js
{
  ...baseTheme,
  buttons: {
    primary: {
      bg: 'primary',
      color: 'white',
    },
    secondary: {
      bg: 'secondary',
      color: 'black',
    }
  }
}
```

A variant can then be applied to a component in the following way:

```
<Box variant="buttons.primary">Box content</Box>
```

This PR will address changes to the following:

- [x] Components
- [ ] Hooks

<a name="qa"></a>

## Acceptance Crtieria

- [x] Flex should render a `div` element with the `display: flex` property on it
- [x] Flex should have all of the flexbox styled-system props available
- [x] Box and Flex should accept the `variant` prop
- [x] Box and Flex should be styled according to the variant defined in the theme

<a name="API Changes"></a>

## API Changes

- [ ] **[Breaking]** This PR introduces breaking changes
- [ ] **[Documentation]** Documentation has been written about the API change

<a name="deployment"></a>

### Before merge to master

**Note**: _Code merged to master should be safe to automatically deploy to production as-is._

- [x] **[QA]** User-testing/quality assurance done
- [x] **[Test]** All components and hooks should have a corresponding unit test
- [x] **[Storybook]** All components and hooks should have a corresponding Storybook story
- [ ] **[Documentation]** Documentation has been written or updated
- [ ] **[Version]** Version number has been updated

### CODE REVIEW

File: packages/box/src/Box.tsx [hunk 1]
```diff
@@ -49,11 +50,20 @@ const shouldForwardProp = createShouldForwardProp([
  *
  * @param props any
  */
-const sx = (props: any): SerializedStyles => {
+function sx(props: any): SerializedStyles {
   return css(props.sx)(props.theme)
 }
 
-type BoxProps = { as?: React.ElementType } & SpaceProps &
+function variant(props: any) {
```
Comment (alexanderchan) [inline packages/box/src/Box.tsx]: can we use this?
```
type VariantProps = {
  theme: Theme
  variant?: string
}
```
and `{ Theme } from 'theme-ui'`

Comment (wesleycole) [reply packages/box/src/Box.tsx]: I tried that and something was bombing with it but I'll try it again and see if it works.

Comment (alexanderchan) [reply packages/box/src/Box.tsx]: Can we also export this since we'll probably have to re-use it?

Comment (wesleycole) [reply packages/box/src/Box.tsx]: Yep!

Comment (wesleycole) [reply packages/box/src/Box.tsx]: Types are fixed and exported both the interface and the `variant` function.

**Pull request merged**